### PR TITLE
feat: cmd args & more

### PR DESCRIPTION
**Pull request by izturn**
1. upgrade to go v1.21
2.  use built-in replace customize funcs
3. support command arguments
4. change examples' mod name
5. more go-style
6. update go-control-plane to 0.12.0

Signed-off-by: Gang Liu [gang.liu@daocloud.io](mailto:gang.liu@daocloud.io)

### CODE REVIEW

**Comment by izturn** [comment]
@wrossmorrow PTAL

**Comment by wrossmorrow** [comment]
Thanks, sorry for the delay. I'll try to review tonight or tomorrow. I appreciate the request.

File: context.go [hunk 1]
```diff
@@ -29,23 +31,51 @@ type PhaseResponse struct {
...
+		}
+
+		if len(h.Value) > 0 {
+			headers[h.Key] = strings.Split(h.Value, ",")
+		} else {
+			rawBytes[h.Key] = h.RawValue
+		}
+	}
+	return
+}
```
Comment (wrossmorrow) [inline context.go:69]:
```suggestion
}

```

File: context.go [hunk 2]
```diff
@@ -71,7 +101,7 @@ func initReqCtx(rc *RequestContext, headers *corev3.HeaderMap) error {
 			rc.RequestId = h.Value
 
 		default:
-			rc.Headers[h.Key] = strings.Split(h.Value, ",")
+			//placeholder
```
Comment (wrossmorrow) [inline context.go:104]: The purpose of this as opposed to assignment was to filter headers in the context to "standard" headers in specific named context fields, while still storing all headers for arbitrary access. Doesn't this change break that? 

I guess it doesn't really matter other than "replication" and potential "confusion" if someone prints or maybe an error if parsing all headers and not expecting the `envoy` `:***` special headers.

File: context.go [hunk 3]
```diff
@@ -2,7 +2,9 @@ package extproc
 
 import (
 	"errors"
+	"fmt"
 	"log"
+	"slices"
```
Comment (wrossmorrow) [inline context.go:7]: Thanks.

File: examples/data.go [hunk 4]
```diff
@@ -6,42 +6,48 @@ import (
...
-	opts := ep.NewOptions()
-	opts.UpdateExtProcHeader = true
-	opts.UpdateDurationHeader = true
-	return opts
+func (s *dataRequestProcessor) GetOptions() *ep.ProcessingOptions {
+	return s.opts
 }
 
-func (s dataRequestProcessor) ProcessRequestHeaders(ctx *ep.RequestContext, headers map[string][]string) error {
+func (s *dataRequestProcessor) ProcessRequestHeaders(ctx *ep.RequestContext, headers map[string][]string, headerRawValues map[string][]byte) error {
```
Comment (wrossmorrow) [inline examples/data.go:21]: I think this is a place where passing a struct with string and byte fields is preferable; I'll make a branch with specific suggestions to work this idea out.

File: examples/go.mod [hunk 5]
```diff
@@ -1,6 +1,6 @@
-module github.com/wrossmorrow/envoy-extproc-sdk-go/examples/timer
+module github.com/wrossmorrow/envoy-extproc-sdk-go/examples
 
-go 1.19
+go 1.21
```
Comment (wrossmorrow) [inline examples/go.mod:3]: Thanks

File: examples/dedup.go [hunk 6]
```diff
@@ -119,6 +118,13 @@ func (s dedupRequestProcessor) ProcessResponseBody(ctx *ep.RequestContext, body
 	return ctx.ContinueRequest()
 }
 
-func (s dedupRequestProcessor) ProcessResponseTrailers(ctx *ep.RequestContext, trailers map[string][]string) error {
+func (s *dedupRequestProcessor) ProcessResponseTrailers(ctx *ep.RequestContext, trailers map[string][]string, rawValues map[string][]byte) error {
 	return ctx.ContinueRequest()
 }
+
+func (s *dedupRequestProcessor) Init(opts *ep.ProcessingOptions, extnonFlagArgsraArgs []string) error {
```
Comment (wrossmorrow) [inline examples/dedup.go:125]:
```suggestion
func (s *dedupRequestProcessor) Init(opts *ep.ProcessingOptions, nonFlagArgs []string) error {
```

**Review by wrossmorrow — CHANGES_REQUESTED** [review]
The possibility of byte headers is a good observation, I think I'd prefer encapsulation and am working on specific suggestions. I've noticed a few other things as well that may get cleaned up (on me, not these changes).

**Comment by wrossmorrow** [comment]
See here: https://github.com/wrossmorrow/envoy-extproc-sdk-go/pull/3

**Comment by wrossmorrow** [comment]
Also there is a [Java SDK variant being used](https://github.com/wrossmorrow/envoy-extproc-sdk-java), and that is a place where more ideas might be getting fleshed out.

File: examples/data.go (hunk 4 above)
Comment (izturn) [reply examples/data.go:21]: with `*`, is to make lint happy

**Comment by izturn** [comment]
i am ok with #3, let's move on with it

**Pull request merged**